### PR TITLE
Speed top-p sampler up by only sampling from top-k tokens

### LINKED ISSUE #963: Investigate top-p performance
Currently our top-p (nucleus) sampler has the worst performance of our entire sampling offering, likely because of the sort operation we need to do for our softmax probabilities. See [these rough numbers](https://github.com/keras-team/keras-nlp/commit/41af01200c5d9898f381477b6a67a87b6951f844#diff-576c27794bae284efc278336a16d0649aaf599973a64d4b65dc67bc99bb7f000) for example, where top-p is slower than beam, despite needing much less computation on the model itself.

We should see if there is anything we can do to speed up top-p sampling with XLA (e.g. a more XLA friendly sort op).

### PR DESCRIPTION
**Pull request by chenmoneygithub**
resolve #963

### CODE REVIEW

**Comment by chenmoneygithub** [comment]
Tested on CPU, with 25 iterations, after the fix it's 37s compared to 55s.

File: keras_nlp/samplers/top_p_sampler.py [hunk 1]
```diff
@@ -34,6 +34,8 @@ class TopPSampler(Sampler):
 
     Args:
         p: float, the `p` value of top-p.
+        sample_from_top_k: int, defaults to None. If set, only sample from top
```
Comment (mattdangerw) [inline keras_nlp/samplers/top_p_sampler.py:37]: I would just call this `k` honestly (parallel with the same option in top-k), and in the description  that it is a heuristic cutoff point.

File: keras_nlp/samplers/top_p_sampler.py [hunk 2]
```diff
@@ -65,17 +67,28 @@ def next(prompt, cache, index):
     def __init__(
         self,
         p=0.1,
+        sample_from_top_k=40,
```
Comment (mattdangerw) [inline keras_nlp/samplers/top_p_sampler.py:70]: I think this does need to default to `None` for now. We could try to set a default value here, but I think it would need to depend on both `p` and the vocabulary size, and be a somewhat interesting research question to set right. Let's default to None and leave this as end user config for now.

File: keras_nlp/samplers/top_p_sampler.py [hunk 3]
```diff
@@ -65,17 +67,28 @@ def next(prompt, cache, index):
...
         self.p = p
+        self.sample_from_top_k = sample_from_top_k
         self.seed = seed
 
     def get_next_token(self, probabilities):
-        # Sort preds in descending order.
-        sorted_preds, sorted_indices = tf.math.top_k(
-            probabilities, k=tf.shape(probabilities)[1], sorted=True
-        )
+        if self.sample_from_top_k is not None:
```
Comment (mattdangerw) [inline keras_nlp/samplers/top_p_sampler.py:79]: This can be simpler.

```
cutoff = tf.shape(probabilities)[1]
if self.k is not None:
    cutoff = tf.math.minimum(cutoff, self.k)
sorted_preds, sorted_indices = tf.math.top_k(probabilities, k=cutoff, sorted=True)

File: keras_nlp/samplers/top_p_sampler_test.py [hunk 4]
```diff
@@ -88,6 +88,28 @@ def test_early_stopping(self):
         )
         self.assertEqual(self.join_as_string(output), ["sequentzzzzz"])
 
+    def test_sample_from_all_tokens(self):
```
Comment (mattdangerw) [inline keras_nlp/samplers/top_p_sampler_test.py:91]: I don't think this test adds to much. If we want to add a test here, it should test that our cutoff point is working.

e.g. `p=1.0`, `k=5`, uniform logits, assert that all outputs in the first five vocab options.

File: keras_nlp/samplers/top_p_sampler.py (hunk 1 above)
Comment (chenmoneygithub) [reply keras_nlp/samplers/top_p_sampler.py:37]: My thought is having both `p` and `k` is a bit confusing to readers, here `p` is the main arg while `k` is a supporting role, using `p` and `k` combo kinda suggest they are parallel to each other.

No strong opinion here and agree `sample_from_top_k` isn't a great name.

File: keras_nlp/samplers/top_p_sampler_test.py (hunk 4 above)
Comment (chenmoneygithub) [reply keras_nlp/samplers/top_p_sampler_test.py:91]: This test case is just a sanity check on sampler call with `k=None` works.

Agree the proposed test works better! changed.

File: keras_nlp/samplers/top_p_sampler.py [hunk 5]
```diff
@@ -34,6 +34,8 @@ class TopPSampler(Sampler):
 
     Args:
         p: float, the `p` value of top-p.
+        k: int, defaults to None. If set, only sample from top `k` tokens.
```
Comment (mattdangerw) [inline keras_nlp/samplers/top_p_sampler.py:37]: ... If set, this argument defines a heuristic "top-k" cutoff applied before the "top-p" sampling. All logits not in the top `k` will be discarded, and the remaining set will be sorted to find a cutoff point for `p`. Setting this argument can significantly speed sampling by reducing the size of the sort.

**Review by mattdangerw — APPROVED** [review]
Looks good! Think we should be more detailed in the docstring for this one.

File: keras_nlp/samplers/top_p_sampler.py (hunk 5 above)
Comment (chenmoneygithub) [reply keras_nlp/samplers/top_p_sampler.py]: thanks!
